File: scripts/parse-events.py
#!/usr/bin/env python3
"""Парсер событий iron-star.com/event/ → .md файлы в src/content/events/"""

import re, os, sys
from datetime import datetime

import requests
from bs4 import BeautifulSoup

def parse_event_text(text: str) -> dict | None:
    """Парсит строку карточки события в структуру."""
    text = text.strip()

    # Паттерн: НАЗВАНИЕ ДД.ММ.ГГГГ ГОРОД [СТАТУС] ДИСТАНЦИИ
    # Статусы: Осталось мало слотов | Продано | (отсутствует = открыта)
    pattern = re.compile(
        r'^(.+?)\s+(\d{2}\.\d{2}\.\d{4})\s+(.+?)\s+'
        r'(Осталось\s+мало\s+слотов|Продано|Лист\s+ожидания)?\s*'
        r'([\d,.]+\s*км(?:\s+[\d,.]+\s*км)*)$',
        re.IGNORECASE
    )
    m = pattern.match(text)
    if not m:
        print(f"  SKIP (no match): {text[:80]}", file=sys.stderr)
        return None

    name = m.group(1).strip()
    date_str = m.group(2).strip()
    city = m.group(3).strip()
    status_raw = (m.group(4) or '').strip()
    distances_str = m.group(5).strip()

    # Нормализуем статус
    status_map = {
        'осталось мало слотов': 'low',
        'продано': 'sold_out',
        'лист ожидания': 'waitlist',
    }
    status = status_map.get(status_raw.lower(), 'open')

    # Парсим дату
    date = datetime.strptime(date_str, '%d.%m.%Y')

    # Парсим дистанции
    dist_pattern = re.compile(r'([\d,]+)\s*км')
    distances = [float(d.replace(',', '.')) for d in dist_pattern.findall(distances_str)]

    # Определяем sportType по количеству и порядку дистанций
    if len(distances) == 1:
        # Одна дистанция: бег или плавание. По названию определяем.
        name_upper = name.upper()
        if any(kw in name_upper for kw in ['SWIM', 'ПЛАВ']):
            sport_type = 'swim'
            mapped = {'swim': distances[0]}
        else:
            sport_type = 'run'
            mapped = {'run': distances[0]}
    elif len(distances) == 2:
        # Две дистанции: swimrun или акватлон или детский дуатлон
        name_upper = name.upper()
        if 'SWIMRUN' in name_upper or 'СВИМРАН' in name_upper:
            sport_type = 'swimrun'
            mapped = {'swim': distances[0], 'run': distances[1]}
        elif 'AQUATHLON' in name_upper or 'АКВАТЛОН' in name_upper:
            sport_type = 'aquathlon'
            mapped = {'swim': distances[0], 'run': distances[1]}
        else:
            # Детские старты: swim + run (иногда bike пропущен)
            sport_type = 'aquathlon'
            mapped = {'swim': distances[0], 'run': distances[1]}
    elif len(distances) == 3:
        sport_type = 'triathlon'
        mapped = {'swim': distances[0], 'bike': distances[1], 'run': distances[2]}
    else:
        print(f"  SKIP: unexpected distance count {len(distances)}: {text[:80]}", file=sys.stderr)
        return None

    return {
        'title': name,
        'date': date,
        'city': city,
        'sportType': sport_type,
        'status': status,
        'distances': mapped,
    }


def fetch_events() -> list[str]:
    """Получает список текстов карточек событий с iron-star.com/event/"""
    url = 'https://iron-star.com/event/'
    headers = {
        'User-Agent': 'Mozilla/5.0 (compatible; PlanExBot/1.0; +https://dimino.me)',
    }
    resp = requests.get(url, headers=headers, timeout=30)
    resp.raise_for_status()

    soup = BeautifulSoup(resp.text, 'html.parser')

    # Ищем ссылки, содержащие дату в тексте
    date_pattern = re.compile(r'\d{2}\.\d{2}\.\d{4}')

    events = []
    seen = set()
    for a in soup.find_all('a', href=True):
        text = a.get_text(separator=' ', strip=True)
        if not text:
            continue
        # Пропускаем кнопки и системные ссылки
        if text.lower() in ('регистрация', 'подробнее', 'лист ожидания', 'eng', 'rus',
                            'фильтр соревнований', 'личный кабинет', 'архив результатов',
                            'помощь', 'подписаться на рассылку'):
            continue
        # Пропускаем дубликаты
        if text in seen:
            continue
        # Проверяем, что текст содержит дату и "км"
        if date_pattern.search(text) and 'км' in text.lower():
            seen.add(text)
            events.append(text)

    return events


def slugify(title: str, date: datetime) -> str:
    """Генерирует slug для имени файла."""
    name_slug = re.sub(r'[^a-zа-яё0-9]+', '-', title.lower()).strip('-')
    date_slug = date.strftime('%Y-%m-%d')
    return f'{date_slug}-{name_slug}'


def generate_md(event: dict) -> str:
    """Генерирует содержимое .md файла."""
    distances = event['distances']
    dist_lines = '\n'.join(f'    {k}: {v}' for k, v in distances.items())

    # Формируем тело без заголовка h1 (данные уже в frontmatter)
    body = ''
    for k, v in distances.items():
        label = {'swim': '🏊 Плавание', 'bike': '🚴 Вело', 'run': '🏃 Бег'}.get(k, k)
        body += f'- {label}: {v} км\n'

    return f'''---
title: "{event['title']}"
date: {event['date'].strftime('%Y-%m-%d')}
city: "{event['city']}"
sportType: {event['sportType']}
status: {event['status']}
distances:
{dist_lines}
---

**Дата:** {event['date'].strftime('%d.%m.%Y')}  
**Город:** {event['city']}  
**Тип:** {event['sportType']}  
**Статус слотов:** {event['status']}

## Дистанции

{body}
[Регистрация на iron-star.com](https://iron-star.com/event/)
'''


def main():
    script_dir = os.path.dirname(os.path.abspath(__file__))
    project_dir = os.path.dirname(script_dir)
    events_dir = os.path.join(project_dir, 'src', 'content', 'events')

    os.makedirs(events_dir, exist_ok=True)

    print('Fetching events from iron-star.com/event/ ...')
    texts = fetch_events()
    print(f'Found {len(texts)} event cards')

    parsed = 0
    saved = 0

    for text in texts:
        event = parse_event_text(text)
        if not event:
            continue
        parsed += 1

        slug = slugify(event['title'], event['date'])
        md_content = generate_md(event)
        filepath = os.path.join(events_dir, f'{slug}.md')

        with open(filepath, 'w', encoding='utf-8') as f:
            f.write(md_content)
        print(f'  ✓ {slug}.md')
        saved += 1

    print(f'\nDone: {parsed} parsed, {saved} saved to {events_dir}')


if __name__ == '__main__':
    main()

File: scripts/parse-events.mjs
#!/usr/bin/env node
/**
 * Парсер событий iron-star.com/event/ → .md файлы в src/content/events/
 * Использует Playwright, так как сайт — Nuxt SPA.
 *
 * Usage: node scripts/parse-events.mjs
 */

import { chromium } from 'playwright';
import { readFile, writeFile, mkdir } from 'node:fs/promises';
import { existsSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const PROJECT_DIR = join(__dirname, '..');
const EVENTS_DIR = join(PROJECT_DIR, 'src', 'content', 'events');

/**
 * Определяет sportType по названию и дистанциям.
 */
function detectSportType(title, dists) {
  const nameUp = title.toUpperCase();

  if (dists.length === 1) {
    if (/SWIM|ПЛАВ/.test(nameUp)) return { type: 'swim', dists: { swim: dists[0] } };
    if (/RUN|БЕГ|IRONLADY|MANSTAR|STARKIDS|5K/.test(nameUp))
      return { type: 'run', dists: { run: dists[0] } };
    return { type: 'run', dists: { run: dists[0] } };
  }

  if (dists.length === 2) {
    if (/SWIMRUN|СВИМРАН/.test(nameUp))
      return { type: 'swimrun', dists: { swim: dists[0], run: dists[1] } };
    if (/AQUATHLON|АКВАТЛОН/.test(nameUp))
      return { type: 'aquathlon', dists: { swim: dists[0], run: dists[1] } };
    // Детские старты, relay — swim/run
    if (/RELAY/.test(nameUp)) return { type: 'swim', dists: { swim: dists[0] } };
    return { type: 'aquathlon', dists: { swim: dists[0], run: dists[1] } };
  }

  if (dists.length === 3) {
    return { type: 'triathlon', dists: { swim: dists[0], bike: dists[1], run: dists[2] } };
  }

  return null;
}

/**
 * Определяет статус слотов.
 */
function detectStatus(statusText) {
  const s = (statusText || '').toLowerCase();
  if (s.includes('мало')) return 'low';
  if (s.includes('продано')) return 'sold_out';
  if (s.includes('ожидани')) return 'waitlist';
  return 'open';
}

/**
 * Парсит дистанцию из строки вида "1,93 км" или "1.93 км"
 */
function parseDistance(text) {
  const cleaned = text.replace(/\s*км\s*/i, '').replace(',', '.');
  const val = parseFloat(cleaned);
  return isNaN(val) ? 0 : val;
}

/**
 * Slug для имени файла
 */
function slugify(title, dateStr) {
  const nameSlug = title
    .toLowerCase()
    .replace(/[^a-zа-яё0-9]+/g, '-')
    .replace(/^-|-$/g, '');
  return `${dateStr}-${nameSlug}`;
}

/**
 * Генерирует .md контент
 */
function generateMd(event) {
  const { title, dateStr, city, sportType, status, distances } = event;
  const distKeys = Object.entries(distances);

  const frontmatterDists = distKeys.map(([k, v]) => `    ${k}: ${v}`).join('\n');

  const bodyLines = distKeys
    .map(([k, v]) => {
      const labels = { swim: '🏊 Плавание', bike: '🚴 Вело', run: '🏃 Бег' };
      return `- ${labels[k] || k}: ${v} км`;
    })
    .join('\n');

  return `---
title: "${title}"
date: ${dateStr}
city: "${city}"
sportType: ${sportType}
status: ${status}
distances:
${frontmatterDists}
---

**Дата:** ${dateStr.replace(/-/g, '.')}
**Город:** ${city}
**Тип:** ${sportType}
**Статус слотов:** ${status}

## Дистанции

${bodyLines}

[Регистрация на iron-star.com](https://iron-star.com/event/)
`;
}

async function main() {
  await mkdir(EVENTS_DIR, { recursive: true });

  console.log('Launching browser...');
  const browser = await chromium.launch({ headless: true });
  const page = await browser.newPage();

  try {
    console.log('Navigating to iron-star.com/event/ ...');
    await page.goto('https://iron-star.com/event/', { waitUntil: 'networkidle', timeout: 30000 });

    // Ждём загрузки карточек
    await page.waitForSelector('.event-item-wrap', { timeout: 10000 });

    // Скроллим до конца для загрузки всех событий (виртуальный скроллинг)
    let prevCount = 0;
    for (let i = 0; i < 20; i++) {
      await page.evaluate(() => window.scrollTo(0, document.body.scrollHeight));
      await page.waitForTimeout(500);
      const count = await page.$$eval('.event-item-wrap', (els) => els.length);
      if (count === prevCount && count > 0) break;
      prevCount = count;
    }

    // Извлекаем данные
    const events = await page.$$eval('.event-item-wrap', (els) =>
      els.map((el) => {
        const link = el.querySelector('a.event-item');
        const title = el.querySelector('.title')?.textContent?.trim() || '';
        const date = el.querySelector('.date')?.textContent?.trim() || '';
        const place = el.querySelector('.place')?.textContent?.trim() || '';
        // Статус в div внутри .event-image
        const imageDiv = el.querySelector('.event-image');
        const statusDiv = imageDiv?.querySelector('div');
        const statusText = statusDiv?.textContent?.trim() || '';
        const distances = Array.from(el.querySelectorAll('.distance')).map((d) =>
          d.textContent.trim(),
        );
        return { title, date, place, statusText, distances, href: link?.href || '' };
      }),
    );

    console.log(`Found ${events.length} event cards\n`);

    let saved = 0;
    let skipped = 0;

    for (const raw of events) {
      const { title, date: dateRaw, place: city, statusText, distances: distStrs, href } = raw;

      // Парсим дату ДД.ММ.ГГГГ → YYYY-MM-DD
      const dateMatch = dateRaw.match(/^(\d{2})\.(\d{2})\.(\d{4})$/);
      if (!dateMatch) {
        console.log(`  SKIP (bad date): ${title} ${dateRaw}`);
        skipped++;
        continue;
      }
      const dateStr = `${dateMatch[3]}-${dateMatch[1]}-${dateMatch[2]}`;

      // Парсим дистанции
      const dists = distStrs.map(parseDistance).filter((d) => d > 0);
      if (dists.length === 0) {
        console.log(`  SKIP (no distances): ${title}`);
        skipped++;
        continue;
      }

      const sport = detectSportType(title, dists);
      if (!sport) {
        console.log(`  SKIP (unknown sport): ${title} ${dists.length} dists`);
        skipped++;
        continue;
      }

      const status = detectStatus(statusText);
      const slug = slugify(title, dateStr);
      const md = generateMd({
        title,
        dateStr,
        city,
        sportType: sport.type,
        status,
        distances: sport.dists,
      });

      const filepath = join(EVENTS_DIR, `${slug}.md`);
      await writeFile(filepath, md, 'utf-8');
      console.log(`  ✓ ${slug}.md (${sport.type}, ${status})`);
      saved++;
    }

    console.log(`\nDone: ${saved} saved, ${skipped} skipped → ${EVENTS_DIR}`);
  } finally {
    await browser.close();
  }
}

main().catch((err) => {
  console.error('FATAL:', err);
  process.exit(1);
});
